spec/integration/federation/shared_receive_relayable.rb

```diff
@@ -0,0 +1,94 @@
+                                              .and_call_original
+
+    Workers::ReceiveEncryptedSalmon.new.perform(user.id, generate_xml(entity, remote_user, user))
+    expect(klass.exists?(guid: entity.guid)).to be(false)
+  end
+
+  it "treats downstream receive correctly" do
+    entity = FactoryGirl.build(
+      entity_name,
+      parent_guid: remote_message.guid,
+      diaspora_id: remote_user.person.diaspora_handle
+    )
+
+    expect(DiasporaFederation.callbacks).to receive(:trigger)
+                                              .with(:fetch_private_key_by_id, remote_user.person.diaspora_handle)
+                                              .once
+                                              .and_return(remote_user.encryption_key)
+    expect(DiasporaFederation.callbacks).to receive(:trigger)
+                                              .with(:fetch_private_key_by_post_guid, remote_message.guid)
+                                              .once
+                                              .and_return(remote_user.encryption_key)
+
+    Workers::ReceiveEncryptedSalmon.new.perform(user.id, generate_xml(entity, remote_user, user))
+    received_entity = klass.find_by(guid: entity.guid)
+    expect(received_entity).not_to be_nil
+    expect(received_entity.author.diaspora_handle).to eq(remote_user.person.diaspora_handle)
+  end
+
+  it "rejects a downstream entity with a malformed author signature" do
+    entity = FactoryGirl.build(
+      entity_name,
+      parent_guid:      local_message.guid,
+      diaspora_id:      remote_user.person.diaspora_handle,
+      author_signature: "I'm a wrong signature"
+    )
+
+    expect(DiasporaFederation.callbacks).to receive(:trigger)
+                                              .with(:fetch_private_key_by_post_guid, local_message.guid)
+                                              .once
+                                              .and_return(remote_user.encryption_key)
+
+    Workers::ReceiveEncryptedSalmon.new.perform(user.id, generate_xml(entity, remote_user, user))
+    expect(klass.exists?(guid: entity.guid)).to be(false)
+  end
+
+  it "declines downstream receive when parent signature is wrong" do
+    entity = FactoryGirl.build(
+      entity_name,
+      parent_guid:             remote_message.guid,
+      diaspora_id:             remote_user.person.diaspora_handle,
+      parent_author_signature: "I'm a wrong signature"
+    )
+
+    expect(DiasporaFederation.callbacks).to receive(:trigger)
+                                              .with(:fetch_private_key_by_id, remote_user.person.diaspora_handle)
+                                              .once
+                                              .and_return(remote_user.encryption_key)
+
+    Workers::ReceiveEncryptedSalmon.new.perform(user.id, generate_xml(entity, remote_user, user))
+    expect(klass.exists?(guid: entity.guid)).to be(false)
```

Comment:
When I try this instead I get [the following](https://gist.github.com/cmrd-senya/1c1b1d880682ae471b88).